internal/rpc/starknet/types_test.go

```diff
@@ -0,0 +1,276 @@
+package starknet
+
+import (
+	"errors"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/NethermindEth/juno/pkg/felt"
+	gocmp "github.com/google/go-cmp/cmp"
+
+	"gotest.tools/assert"
+)
+
+func TestBlockId_UnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		name string
+		data []byte
+		want BlockId
+		err  error
+	}{
+		{
+			name: "empty data",
+			data: []byte{},
+			want: BlockId{},
+			err:  io.EOF,
+		},
+		{
+			name: "empty string",
+			data: []byte(`""`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "negative number",
+			data: []byte(`-1`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "float number",
+			data: []byte(`1.1`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "zero string",
+			data: []byte(`"0"`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "zero number",
+			data: []byte(`0`),
+			want: BlockId{
+				value:  int64(0),
+				idType: blockIdNumber,
+			},
+			err: nil,
+		},
+		{
+			name: "positive number",
+			data: []byte(`1`),
+			want: BlockId{
+				value:  int64(1),
+				idType: blockIdNumber,
+			},
+			err: nil,
+		},
+		{
+			name: "pending tag",
+			data: []byte(`"pending"`),
+			want: BlockId{
+				idType: blockIdTag,
+				value:  "pending",
+			},
+			err: nil,
+		},
+		{
+			name: "latest tag",
+			data: []byte(`"latest"`),
+			want: BlockId{
+				idType: blockIdTag,
+				value:  "latest",
+			},
+			err: nil,
+		},
+		{
+			name: "block hash",
+			data: []byte(`"0x07d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"`),
+			want: BlockId{
+				idType: blockIdHash,
+				value:  new(felt.Felt).SetHex("0x07d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"),
+			},
+			err: nil,
+		},
+		{
+			name: "invalid block hash",
+			data: []byte(`"0x7d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "unexpected JSON type",
+			data: []byte(`{}`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var got BlockId
+			if err := got.UnmarshalJSON(tt.data); !errors.Is(err, tt.err) {
+				t.Errorf("BlockId.UnmarshalJSON() error = %v, wantErr %v", err, tt.err)
```

Comment:
For the sake of consistency with how all the other failed tests are reported. 

```suggestion
				t.Errorf("BlockId.UnmarshalJSON() = %v, want %v", err, tt.err)
```